examples/BERT/cross_lingual_mlm_task.py

```diff
@@ -116,40 +121,87 @@ def text_transform(x: str) -> List:
     pad_id = vocab(['pad'])[0]
     ntokens = len(vocab)
 
+    xlmr = XLMRModelShards(ntokens, args.emsize, args.nhead, args.nhid, args.dropout)
+    mlm = MLMShards(ntokens, args.emsize)
+    devices = [f"cuda:{i}" for i in range(args.gpus)] if torch.cuda.is_available() else ["cpu"]
+
+    if len(devices) == 1:
+        shards = [nn.Sequential(
+            xlmr.xlmr_embed(),
+            xlmr.encoder_layers(args.nlayers),
+            mlm.mlm()
+        )]
+    elif len(devices) == 2:
+        assert args.nlayers % 2 == 0
+        shards = [
+            nn.Sequential(
+                xlmr.xlmr_embed(),
+                xlmr.encoder_layers(args.nlayers // 2)
+            ),
+            nn.Sequential(
+                xlmr.encoder_layers(args.nlayers // 2),
+                mlm.mlm()
+            )
+        ]
+    else:
+        encoder_gpus = (args.gpus - 2)
+        assert args.nlayers % encoder_gpus == 0
+        encoders_per_gpu = args.nlayers // encoder_gpus
+        shards = [
+            xlmr.xlmr_embed(),
+            *[xlmr.encoder_layers(encoders_per_gpu) for _ in range(encoder_gpus)],
+            mlm.mlm()
+        ]
+
+    print('Shards parameters:')
+    total = 0
+    for i, shard in enumerate(shards):
+        params = count_model_param(shard)
+        total += params
+        print(f'shard{i} = {int(params)}M')
+    print(f'total = {int(total)}M')
+
+    print("Allocating memory")
     if not args.dist:
-        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
-        model = CrossLingualMLMTask(ntokens, args.emsize, args.nhead, args.nhid, args.nlayers, args.dropout)
-        model = model.to(device)
+        model = LocalPipeline(shards, devices)
```

Comment:
Is this the situation like DDP? If so, should I set `args.dist` to `False`? But I thought we consider `DDP` under distributed. Just curious.

Reply:
Sorry for the wrong naming. Maybe it makes sense to:
1. Rename `LocalPipeline` to `SingleProcessPipeline` and `DistPipeline` to `RPCPipeline`
2. `RPCPipeline` will have two implementations CPU RPC and CUDA RPC
3. Replace `args.dist` and `args.cuda_rpc` with 
a) no argument to use `SingleProcessPipeline`
b) `args.use_cpu_rpc` to use `RPCPipeline` with CPU RPC
c) `args.use_cuda_rpc`  to use `RPCPipeline` with CUDA RPC
4. TBD: `LocalPipeline` can utilize DDP. I can implement it in the next pull request

WDYT?